crawl.go

```diff
@@ -0,0 +1,198 @@
+	"/dns4/bootstrap-0.starpool.in/tcp/12757/p2p/12D3KooWGHpBMeZbestVEWkfdnC9u7p6uFHXL1n7m1ZBqsEmiUzz",
+	"/dns4/bootstrap-1.starpool.in/tcp/12757/p2p/12D3KooWQZrGH1PxSNZPum99M1zNvjNFM33d1AAu5DcvdHptuU7u",
+	"/dns4/node.glif.io/tcp/1235/p2p/12D3KooWBF8cpp65hp2u9LK5mh19x67ftAam84z9LsfaquTDSBpt",
+	"/dns4/bootstrap-0.ipfsmain.cn/tcp/34721/p2p/12D3KooWQnwEGNqcM2nAcPtRR9rAX8Hrg4k9kJLCHoTR5chJfz6d",
+	"/dns4/bootstrap-1.ipfsmain.cn/tcp/34723/p2p/12D3KooWMKxMkD5DMpSWsW7dBddKxKT7L2GgbNuckz9otxvkvByP",
+}
+
+var CalibnetPeers = []string{
+	"/dns4/bootstrap-0.calibration.fildev.network/tcp/1347/p2p/12D3KooWJkikQQkxS58spo76BYzFt4fotaT5NpV2zngvrqm4u5ow",
+	"/dns4/bootstrap-1.calibration.fildev.network/tcp/1347/p2p/12D3KooWLce5FDHR4EX4CrYavphA5xS3uDsX6aoowXh5tzDUxJav",
+	"/dns4/bootstrap-2.calibration.fildev.network/tcp/1347/p2p/12D3KooWA9hFfQG9GjP6bHeuQQbMD3FDtZLdW1NayxKXUT26PQZu",
+	"/dns4/bootstrap-3.calibration.fildev.network/tcp/1347/p2p/12D3KooWMHDi3LVTFG8Szqogt7RkNXvonbQYqSazxBx41A5aeuVz",
+}
+
+type FilecoinPeer struct {
+	ctx            context.Context
+	dht            *dht.IpfsDHT
+	host           host.Host
+	bootstrapPeers []peer.AddrInfo
+}
+
+func FilecoinBootstrapPeers(peerList []string) []peer.AddrInfo {
+	if peerList == nil {
+		peerList = MainnetPeers
+	}
+
+	maddrs := make([]multiaddr.Multiaddr, len(peerList))
+
+	for i, bp := range peerList {
+		var err error
+		maddrs[i], err = multiaddr.NewMultiaddr(bp)
+		if err != nil {
+			return nil
+		}
+	}
+
+	peers, _ := peer.AddrInfosFromP2pAddrs(maddrs...)
+
+	return peers
+}
+
+func SetupFilecoinPeer(
+	ctx context.Context,
+	network NetworkName,
+) (*FilecoinPeer, error) {
+	if network == "" {
+		network = Mainnet
+	}
+
+	var pp protocol.ID
+	var bootstrapPeers []peer.AddrInfo
+
+	switch network {
+	case Mainnet:
+		pp = "/fil/kad/testnetnet"
+		bootstrapPeers = FilecoinBootstrapPeers(MainnetPeers)
+	case Calibnet:
+		pp = "/fil/kad/calibnet"
+		bootstrapPeers = FilecoinBootstrapPeers(CalibnetPeers)
+	}
+
+	var ddht *dht.IpfsDHT
+	priv, _, err := crypto.GenerateKeyPair(crypto.RSA, 2048)
+
+	if err != nil {
+		return &FilecoinPeer{}, err
+	}
+
+	libp2pOpts := []libp2p.Option{
+		libp2p.ListenAddrStrings(
+			"/ip4/0.0.0.0/tcp/9000",      // regular tcp connections
+			"/ip4/0.0.0.0/udp/9000/quic", // a UDP endpoint for the QUIC transport
+		),
+		libp2p.Identity(priv),
+		libp2p.Security(libp2ptls.ID, libp2ptls.New),
+		libp2p.Security(noise.ID, noise.New),
+		libp2p.DefaultTransports,
+		libp2p.Routing(func(h host.Host) (routing.PeerRouting, error) {
+			ddht, err := dht.New(
+				ctx,
+				h,
+				dht.Option(dht.Datastore(dsync.MutexWrap(ds.NewMapDatastore()))),
+				dht.Option(dht.NamespacedValidator("pk", record.PublicKeyValidator{})),
+				dht.Option(dht.NamespacedValidator("ipns", ipns.Validator{KeyBook: h.Peerstore()})),
+				dht.Option(dht.Concurrency(10)),
+				dht.Option(dht.Mode(dht.ModeAuto)),
```

Comment:
I think better this:

```suggestion
				dht.Option(dht.Mode(dht.ModeClient)),
```